<a href="https://colab.research.google.com/github/qweliant/DS-Unit-1-Sprint-3-Statistical-Tests-and-Experiments/blob/master/module1-statistics-probability-and-inference/LS_DS_131_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 3 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [47]:
### YOUR CODE STARTS HERE
### YOUR CODE STARTS HERE
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2019-09-17 14:56:13--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data.2’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.03s   

2019-09-17 14:56:13 (593 KB/s) - ‘house-votes-84.data.2’ saved [18171/18171]



In [48]:
# Load Data
df = pd.read_csv('house-votes-84.data', 
                 header=None,
                 names=['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa'])
print(df.shape)

(435, 17)


In [49]:
df = df.replace({'?':np.NaN, 'n':0, 'y':1}) #use this to create bool table

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [50]:
df.describe()

handicapped-infants  water-project  ...   duty-free  south-africa
count           423.000000     387.000000  ...  407.000000    331.000000
mean              0.442080       0.503876  ...    0.427518      0.812689
std               0.497222       0.500632  ...    0.495327      0.390752
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      1.000000
50%               0.000000       1.000000  ...    0.000000      1.000000
75%               1.000000       1.000000  ...    1.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

In [0]:
dem = df[df['party'] == 'democrat']
rep = df[df['party'] == 'republican']

In [52]:
dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [53]:
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [54]:
df.isnull().sum()

party                     0
handicapped-infants      12
water-project            48
budget                   11
physician-fee-freeze     11
el-salvador-aid          15
religious-groups         11
anti-satellite-ban       14
aid-to-contras           15
mx-missile               22
immigration               7
synfuels                 21
education                31
right-to-sue             25
crime                    17
duty-free                28
south-africa            104
dtype: int64

In [0]:
from scipy.stats import ttest_1samp, ttest_ind

In [56]:
ttest_ind(rep['budget'], dem['budget'] ,nan_policy='omit')

Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)

In [0]:
def run_two_sample_test(one, two):
    return ttest_ind(one, two, nan_policy='omit')

In [65]:
ttest = run_two_sample_test(rep['crime'], dem['crime'])
ttest

Ttest_indResult(statistic=16.342085656197696, pvalue=9.952342705606092e-47)

# 1) Null Hypothesis:

$H_a$: $\bar{x}_1 = \bar{x}_2$ - There is the same amount of positive support for bills amongst reps and dems.

# 2) Alternative Hypothesis:

$H_0$: $\bar{x}_1 > \bar{x}_2$  - Mean approval vote rate for democrats is greater than approval votes for republicans

# 3) Confidence Level: 95% or .95

In [77]:
cols = list(rep.columns)
cols.remove('party')
cols

['handicapped-infants',
 'water-project',
 'budget',
 'physician-fee-freeze',
 'el-salvador-aid',
 'religious-groups',
 'anti-satellite-ban',
 'aid-to-contras',
 'mx-missile',
 'immigration',
 'synfuels',
 'education',
 'right-to-sue',
 'crime',
 'duty-free',
 'south-africa']

In [97]:
stats = []
for c in cols:
    print('This is for',c)
    print(run_two_sample_test(rep[c], dem[c]))
    stats.append( run_two_sample_test(rep[c], dem[c]))

This is for handicapped-infants
Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18)
This is for water-project
Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)
This is for budget
Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)
This is for physician-fee-freeze
Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)
This is for el-salvador-aid
Ttest_indResult(statistic=21.13669261173219, pvalue=5.600520111729011e-68)
This is for religious-groups
Ttest_indResult(statistic=9.737575825219457, pvalue=2.3936722520597287e-20)
This is for anti-satellite-ban
Ttest_indResult(statistic=-12.526187929077842, pvalue=8.521033017443867e-31)
This is for aid-to-contras
Ttest_indResult(statistic=-18.052093200819733, pvalue=2.82471841372357e-54)
This is for mx-missile
Ttest_indResult(statistic=-16.437503268542994, pvalue=5.03079265310811e-47)
This is for immigration
Ttest_indResult(statistic=1.7359117329695164,

# 1) Null Hypothesis:

$H_a$: $\bar{x}_1 = \bar{x}_2$ - There is the same amount of positive support for crime.

# 2) Alternative Hypothesis:

$H_0$: $\bar{x}_1 > \bar{x}_2$  - republicans vote at greater rates for crime

# 3) Confidence Level: 95% or .95


# This is for crime
Ttest_indResult(statistic=16.342085656197696,pvalue=9.952342705606092e-47)


p value is non zero

In [99]:
val = 1 - 0.95
print(val)

0.050000000000000044


# pval is less, so we can reject the null hypothesis, and show...

# Republican vote average for yes

In [100]:
rep['crime'].mean()

0.9813664596273292

# Democrat vote average for yes

In [101]:
dem['crime'].mean()

0.35019455252918286